java/statestore/src/main/java/sleeper/statestore/dynamodb/DynamoDBFileInfoStore.java

```diff
@@ -114,23 +122,96 @@ public void addFile(FileInfo fileInfo) throws StateStoreException {
     }
 
     public void addFile(FileInfo fileInfo, long updateTime) throws StateStoreException {
-        Map<String, AttributeValue> itemValues = fileInfoFormat.createRecord(setLastUpdateTime(fileInfo, updateTime));
         try {
             String tableName = tableName(fileInfo);
-            PutItemRequest putItemRequest = new PutItemRequest()
-                    .withItem(itemValues)
-                    .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL)
-                    .withTableName(tableName);
-            PutItemResult putItemResult = dynamoDB.putItem(putItemRequest);
+            List<TransactWriteItem> writes = new ArrayList<>();
+            writes.add(new TransactWriteItem().withPut(
+                    new Put()
+                            .withItem(fileInfoFormat.createRecord(setLastUpdateTime(fileInfo, updateTime)))
+                            .withTableName(tableName)));
+            writes.add(createOrIncrementFileReferenceCount(fileInfo, updateTime));
+            TransactWriteItemsRequest transactWriteItemsRequest = new TransactWriteItemsRequest()
+                    .withTransactItems(writes)
+                    .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL);
+            TransactWriteItemsResult transactWriteItemsResult = dynamoDB.transactWriteItems(transactWriteItemsRequest);
+            List<ConsumedCapacity> consumedCapacity = transactWriteItemsResult.getConsumedCapacity();
+            double totalConsumed = consumedCapacity.stream().mapToDouble(ConsumedCapacity::getCapacityUnits).sum();
             LOGGER.debug("Put file info for file {} to table {}, capacity consumed = {}",
-                    fileInfo.getFilename(), tableName, putItemResult.getConsumedCapacity().getCapacityUnits());
+                    fileInfo.getFilename(), tableName, totalConsumed);
         } catch (ConditionalCheckFailedException | ProvisionedThroughputExceededException | ResourceNotFoundException
                  | ItemCollectionSizeLimitExceededException | TransactionConflictException
                  | RequestLimitExceededException | InternalServerErrorException e) {
             throw new StateStoreException("Exception calling putItem", e);
         }
     }
 
+    private TransactWriteItem createOrIncrementFileReferenceCount(FileInfo fileInfo, long updateTime) {
+        GetItemRequest getItemRequest = new GetItemRequest()
+                .withTableName(fileReferenceCountTableName)
+                .withConsistentRead(stronglyConsistentReads)
+                .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL)
+                .withKey(Map.of(
+                        TABLE_ID, createStringAttribute(sleeperTableId),
+                        FILENAME, createStringAttribute(fileInfo.getFilename())));
+        GetItemResult getItemResult = dynamoDB.getItem(getItemRequest);
```

Comment:
I don't think this is necessary. You can use an expression to increment the count without querying the existing record, and it's necessary to do that to make the update atomic.